dev-docs/source/ISISReflectometryInterface.rst

```diff
@@ -0,0 +1,309 @@
+
+  .. code-block:: c++
+
+    void QtEventView::onUniformEvenChanged(int numberOfSlices) {
+      m_notifyee->notifyUniformSliceCountChanged(numberOfSlices);
+    }
+
+
+Dependency injection
+####################
+
+A simple example of dependency injection is in the use of an :code:`IMessageHandler` interface to provide a service to display messages to the user. These messages must be displayed by a Qt view. Rather than each view having to implement this, we use one object (in this case the :code:`QtMainWindowView`) to implement this functionality and pass it as an :code:`IMessageHandler` to all of the presenters that need it.
+
+- The :code:`IMessageHandler` interface defines the functions for displaying messages:
+
+  .. code-block:: c++
+
+    class IMessageHandler {
+    public:
+      virtual void giveUserCritical(const std::string &prompt,
+                                    const std::string &title) = 0;
+      ...
+    };
+
+- The :code:`QtMainWindowView` implements these:
+
+  .. code-block:: c++
+
+    void QtMainWindowView::giveUserCritical(const std::string &prompt,
+                                            const std::string &title) {
+      QMessageBox::critical(this, QString::fromStdString(title),
+                            QString::fromStdString(prompt), QMessageBox::Ok,
+                            QMessageBox::Ok);
+    }
+
+- The :code:`QtMainWindowView` sets up the concrete instance (actually just a pointer to itself) and passes it in the construction of anything that needs it, e.g. the :code:`RunsPresenter` (in this case using a factory - more about the `Factory pattern`_ below):
+
+  .. code-block:: c++
+
+    auto messageHandler = this;
+    auto makeRunsPresenter = RunsPresenterFactory(..., messageHandler);
+
+- The :code:`RunsPresenter` then has a simple service it can use to display messages without needing to know anything about the :code:`QtMainWindowView`:
+
+  .. code-block:: c++
+
+    m_messageHandler->giveUserInfo("Search field is empty", "Search Issue");
+
+
+Factory pattern
+###############
+
+The :code:`MainWindowPresenter` constructs the child Batch presenters on demand. This prevents us injecting them in its constructor. In order to follow `Dependency inversion`_, we therefore need to use factories to create the child presenters. Let's use the :code:`MainWindow` -> :code:`Batch` -> :code:`Event` components as an example.
+
+- As mentioned, the :code:`QtMainWindowView` is our entry point. This creates (and owns) the :code:`MainWindowPresenter`. It:
+
+  - creates an :code:`EventPresenterFactory`;
+  - passes this to the :code:`BatchPresenterFactory` constructor so it can create the child :code:`EventPresenter` when needed;
+  - passes this to the :code:`MainWindowPresenter` constructor ready for making a Batch when needed.
```

Comment:
"passes this to the :code:`MainWindowPresenter` constructor ready for making a Batch when needed."
to
"passes this to the :code:`MainWindowPresenter` constructor, which accepts a BatchPresenterFactory, ready for making a Batch when needed."